src/GraphQL/GraphQLExtensions.cs

```diff
@@ -107,33 +110,54 @@ public static bool IsOutputType(this Type type)
         /// </summary>
         public static bool IsOutputType(this IGraphType type)
         {
-            var namedType = type.GetNamedType();
+            var (namedType, namedType2) = type.GetNamedTypes();
             return namedType is ScalarGraphType ||
                    namedType is IObjectGraphType ||
                    namedType is IInterfaceGraphType ||
-                   namedType is UnionGraphType;
+                   namedType is UnionGraphType ||
+                   typeof(ScalarGraphType).IsAssignableFrom(namedType2) ||
+                   typeof(IObjectGraphType).IsAssignableFrom(namedType2) ||
+                   typeof(IInterfaceGraphType).IsAssignableFrom(namedType2) ||
+                   typeof(UnionGraphType).IsAssignableFrom(namedType2);
+            ;
         }
 
         /// <summary>
         /// Indicates if the graph type is an input object graph type.
         /// </summary>
         public static bool IsInputObjectType(this IGraphType type)
         {
-            var namedType = type.GetNamedType();
-            return namedType is IInputObjectGraphType;
+            var (namedType, namedType2) = type.GetNamedTypes();
+            return namedType is IInputObjectGraphType ||
+                   typeof(IInputObjectGraphType).IsAssignableFrom(namedType2);
+        }
+
+        internal static bool IsGraphQLTypeReference(this IGraphType type)
+        {
+            var (namedType, _) = type.GetNamedTypes();
+            return namedType is GraphQLTypeReference;
+        }
+
+        internal static (IGraphType resolvedType, Type type) GetNamedTypes(this IGraphType type)
+        {
+            return type switch
+            {
+                NonNullGraphType nonNull => nonNull.ResolvedType != null ? GetNamedTypes(nonNull.ResolvedType) : (null, GetNamedType(nonNull.Type)),
+                ListGraphType list => list.ResolvedType != null ? GetNamedTypes(list.ResolvedType) : (null, GetNamedType(list.Type)),
+                _ => (type, null)
+            };
         }
 
         /// <summary>
         /// Unwraps any list/non-null graph type wrappers from a graph type and returns the base graph type.
         /// </summary>
         public static IGraphType GetNamedType(this IGraphType type)
         {
-            return type switch
-            {
-                NonNullGraphType nonNull => GetNamedType(nonNull.ResolvedType),
-                ListGraphType list => GetNamedType(list.ResolvedType),
-                _ => type
-            };
+            if (type == null)
+                return null;
+
+            var (namedType, _) = type.GetNamedTypes();
+            return namedType ?? throw new NotSupportedException("Please set ResolvedType property before calling this method or call GetNamedType(this Type type) instead");
```

Comment:
A key part of this PR. The `GetNamedType(this IGraphType type)` method was only intended for use by unwrapping other resolved types. The `GetNamedType(this Type type)` method has no problems in this regard. After throwing the exception, I saw that the calling methods needed to be changed. Typically, they didn't need the type itself. Only in one case did I have to actually create a type. See `SchemaTypes`. We had no errors all this time because we did not configure the graph types this way.